From ee27dfe25da724b08aaeb25fa0f514b769f8fe32 Mon Sep 17 00:00:00 2001
From: Daniel Drubin <idaniel.drubin@intel.com>
Date: Tue, 9 Feb 2016 18:56:15 +0200
Subject: [PATCH 08/12] ISH: fix for KW issues

Added check for register address being within range of ISH device (the address
is received from user-mode via char device API)

Change-Id: I5671582c9173d8485adf54981204b8b5ef543472
Signed-off-by: Daniel Drubin <daniel.drubin@intel.com>
Tracked-On: https://jira01.devtools.intel.com/browse/OAM-18668
Tracked-On: https://jira01.devtools.intel.com/browse/OAM-18669
Reviewed-on: https://android.intel.com:443/468148
---
 drivers/misc/intel-ish/hw-ish-regs.h     | 1 +
 drivers/misc/intel-ish/pci-ish.c         | 9 ++++++++-
 drivers/misc/intel-ish/platform-config.h | 2 +-
 3 files changed, 10 insertions(+), 2 deletions(-)

diff --git a/drivers/misc/intel-ish/hw-ish-regs.h b/drivers/misc/intel-ish/hw-ish-regs.h
index e1c50cdd..cd8b3ab 100644
--- a/drivers/misc/intel-ish/hw-ish-regs.h
+++ b/drivers/misc/intel-ish/hw-ish-regs.h
@@ -41,6 +41,7 @@
 /* REMAP2 to enable DMA (D3 RCR) */
 #define	IPC_REG_ISH_RMP2	 (IPC_REG_BASE + 0x368)
 
+#define	IPC_REG_MAX		(IPC_REG_BASE + 0x400)
 /* register bits - HISR */
 
 /* bit corresponds HOST2ISH interrupt in PISR and PIMR registers */
diff --git a/drivers/misc/intel-ish/pci-ish.c b/drivers/misc/intel-ish/pci-ish.c
index 6e149e4..ef19319 100644
--- a/drivers/misc/intel-ish/pci-ish.c
+++ b/drivers/misc/intel-ish/pci-ish.c
@@ -158,10 +158,17 @@ static ssize_t ishdbg_write(struct file *file, const char __user *ubuf,
 	if (rv)
 		return  -EINVAL;
 	if (sscanf(dbg_req_buf, "%s ", cmd) != 1) {
-		dev_err(&heci_pci_device->dev, "[ish-dbg]) sscanf failed\n");
+		dev_err(&heci_pci_device->dev, "[ish-dbg] sscanf failed\n");
 		return  -EINVAL;
 	}
 	sscanf_match = sscanf(dbg_req_buf + 2, "%x %u", &addr, &count);
+
+	if (addr >= IPC_REG_MAX) {
+		dev_err(&heci_pci_device->dev, "[ish-dbg] address %08X "
+			"out of range\n", addr);
+		return	-EINVAL;
+	}
+
 	if (!strcmp(cmd, "d")) {
 		/* Dump values: d <addr> [count] */
 		if (sscanf_match == 1)
diff --git a/drivers/misc/intel-ish/platform-config.h b/drivers/misc/intel-ish/platform-config.h
index 71df6fb..2aeb82e 100644
--- a/drivers/misc/intel-ish/platform-config.h
+++ b/drivers/misc/intel-ish/platform-config.h
@@ -17,7 +17,7 @@
 #define PLATFORM_CONFIG__H
 
 /* Build ID string */
-#define BUILD_ID	"243-reset-flow-release-fix"
+#define BUILD_ID	"245-reset-flow-release-hidcl-init-ishdbg-fixes"
 
 #define	ISH_DEBUG	0
 #if ISH_DEBUG
-- 
1.9.1

